File: services/backend-go/internal/analytics/global_series.go
package analytics

import (
	"context"
	"math"
	"strconv"
	"strings"

	"github.com/jackc/pgx/v5/pgxpool"
	"tickergarden/backend/internal/deployment"
)

type GlobalFlowGroup struct {
	AssetUID               string          `json:"assetUid"`
	Binding                string          `json:"binding"`
	QuoteAsset             string          `json:"quoteAsset"`
	QuoteDecimals          uint8           `json:"quoteDecimals"`
	TradeCount             uint64          `json:"tradeCount"`
	InternalTradeCount     uint64          `json:"internalTradeCount"`
	UnclassifiedTradeCount uint64          `json:"unclassifiedTradeCount"`
	QuoteVolumeRaw         string          `json:"quoteVolumeRaw"`
	InternalQuoteVolumeRaw string          `json:"internalQuoteVolumeRaw"`
	Fees                   []AssetFeeTotal `json:"fees"`
	UnknownFeeTradeCount   uint64          `json:"unknownFeeTradeCount"`
}
type GlobalFlowPoint struct {
	Timestamp uint64            `json:"timestamp"`
	Groups    []GlobalFlowGroup `json:"groups"`
}
type GlobalFlowSeries struct {
	Coverage    RangeCoverage     `json:"coverage"`
	Interval    uint64            `json:"interval"`
	VolumeBasis string            `json:"volumeBasis"`
	EmptyPolicy string            `json:"emptyPolicy"`
	Points      []GlobalFlowPoint `json:"points"`
}

func validGlobalSeriesRange(from, to, interval uint64) bool {
	switch interval {
	case 60, 300, 900, 3600, 14400, 86400:
	default:
		return false
	}
	return from < to && to <= math.MaxInt64 && from%interval == 0 && to%interval == 0 && (to-from)/interval <= 2000
}

// BuildGlobalFlowSeries uses complete authenticated market inputs for the whole
// range. No balance, reserve, historical market count or price is inferred.
func BuildGlobalFlowSeries(chain uint64, coverage RangeCoverage, interval uint64, inputs []AssetMarketTrades) (GlobalFlowSeries, error) {
	fail := func() (GlobalFlowSeries, error) { return GlobalFlowSeries{}, ErrAssetStats }
	if chain == 0 || !validGlobalSeriesRange(coverage.From, coverage.To, interval) {
		return fail()
	}
	buckets := int((coverage.To - coverage.From) / interval)
	if len(inputs) > 1000 || len(inputs)*buckets > 100000 {
		return fail()
	}
	if _, err := AggregateAssetTrades(chain, coverage, inputs); err != nil {
		return fail()
	}
	partition := make([][][]TradeActivity, len(inputs))
	for i, m := range inputs {
		partition[i] = make([][]TradeActivity, buckets)
		for _, t := range m.Trades.Items {
			ts, _ := strconv.ParseUint(t.Timestamp, 10, 64)
			b := (ts - coverage.From) / interval
			partition[i][b] = append(partition[i][b], t)
		}
	}
	out := GlobalFlowSeries{Coverage: coverage, Interval: interval, VolumeBasis: "CURVE_EXCLUDING_FEE_TAX_OR_POOL_CORE", EmptyPolicy: "ZERO_FLOW_NO_SYNTHETIC_EXECUTIONS", Points: make([]GlobalFlowPoint, 0, buckets)}
	zero := "0x" + strings.Repeat("0", 64)
	for b := 0; b < buckets; b++ {
		from := coverage.From + uint64(b)*interval
		rangeCoverage := coverage
		rangeCoverage.From = from
		rangeCoverage.To = from + interval
		selected := make([]AssetMarketTrades, len(inputs))
		for i, m := range inputs {
			selected[i] = m
			selected[i].Trades.Coverage = rangeCoverage
			selected[i].Trades.Items = partition[i][b]
		}
		groups, err := AggregateAssetTrades(chain, rangeCoverage, selected)
		if err != nil {
			return fail()
		}
		point := GlobalFlowPoint{Timestamp: from, Groups: []GlobalFlowGroup{}}
		for _, g := range groups {
			binding := "registered_stock"
			if g.AssetUID == zero {
				binding = "unbound"
			}
			point.Groups = append(point.Groups, GlobalFlowGroup{g.AssetUID, binding, g.QuoteAsset, g.QuoteDecimals, g.TradeCount, g.InternalTradeCount, g.UnclassifiedTradeCount, g.QuoteVolumeRaw, g.InternalQuoteVolumeRaw, g.Fees, g.UnknownFeeTradeCount})
		}
		out.Points = append(out.Points, point)
	}
	return out, nil
}
func LoadGlobalFlowSeries(ctx context.Context, pool *pgxpool.Pool, m deployment.Manifest, from, to, interval uint64) (GlobalFlowSeries, error) {
	if !validGlobalSeriesRange(from, to, interval) {
		return GlobalFlowSeries{}, ErrAssetStats
	}
	stats, inputs, err := loadGlobalSnapshot(ctx, pool, m, from, to)
	if err != nil {
		return GlobalFlowSeries{}, err
	}
	return BuildGlobalFlowSeries(m.ChainID, stats.Coverage, interval, inputs)
}

File: apps/web/src/v1/globalSeries.ts
import type {GlobalFlowSeriesResponse} from './generated/read-api.ts';
const hash=(v:unknown)=>typeof v==='string'&&/^0x[0-9a-f]{64}$/.test(v);
const address=(v:unknown)=>typeof v==='string'&&/^0x[0-9a-f]{40}$/.test(v);
const count=(v:unknown)=>typeof v==='number'&&Number.isSafeInteger(v)&&v>=0;
const amount=(v:unknown)=>typeof v==='string'&&/^(0|[1-9][0-9]{0,99})$/.test(v);
export function validateGlobalSeries(value:unknown,chain:number,from:number,to:number):GlobalFlowSeriesResponse{
 const p=value as GlobalFlowSeriesResponse;const fail=():never=>{throw new Error('Inconsistent flow series');};
 if(!p||p.chainId!==chain||p.displayOnly!==true||p.interval!==3600||to-from!==86400||from%3600!==0||p.volumeBasis!=='CURVE_EXCLUDING_FEE_TAX_OR_POOL_CORE'||p.emptyPolicy!=='ZERO_FLOW_NO_SYNTHETIC_EXECUTIONS'||!p.coverage)fail();
 const c=p.coverage;if(c.from!==from||c.to!==to||![c.anchorNumber,c.throughNumber,c.projectionNumber].every(count)||c.anchorNumber>=c.throughNumber||c.throughNumber>c.projectionNumber||![c.anchorHash,c.throughHash,c.projectionHash].every(hash)||!Array.isArray(p.points)||p.points.length!==24)fail();
 let identity:string|null=null;const decimals=new Map<string,number>();
 for(const [i,point] of p.points.entries()){
  if(!point||point.timestamp!==from+i*3600||!Array.isArray(point.groups)||point.groups.length>1000)fail();
  const keys:string[]=[];const unique=new Set<string>();
  for(const g of point.groups){
   if(!g||!hash(g.assetUid)||!address(g.quoteAsset)||!count(g.quoteDecimals)||g.quoteDecimals<6||g.quoteDecimals>18||(g.quoteAsset==='0x'+'0'.repeat(40)&&g.quoteDecimals!==18))fail();
   const unbound=g.assetUid==='0x'+'0'.repeat(64);if(g.binding!==(unbound?'unbound':'registered_stock'))fail();
   const key=g.assetUid+':'+g.quoteAsset;if(unique.has(key))fail();unique.add(key);keys.push(key+':'+g.quoteDecimals);
   const old=decimals.get(g.quoteAsset);if(old!==undefined&&old!==g.quoteDecimals)fail();decimals.set(g.quoteAsset,g.quoteDecimals);
   if(![g.tradeCount,g.internalTradeCount,g.unclassifiedTradeCount,g.unknownFeeTradeCount].every(count)||g.tradeCount!==g.internalTradeCount+g.unclassifiedTradeCount||g.unknownFeeTradeCount>g.tradeCount||!amount(g.quoteVolumeRaw)||!amount(g.internalQuoteVolumeRaw)||BigInt(g.internalQuoteVolumeRaw)>BigInt(g.quoteVolumeRaw)||(g.internalTradeCount===0&&g.internalQuoteVolumeRaw!=='0')||(g.tradeCount===0&&g.quoteVolumeRaw!=='0')||(g.tradeCount>0&&g.quoteVolumeRaw==='0')||!Array.isArray(g.fees)||g.fees.length>2000||(g.tradeCount===0&&g.fees.length!==0))fail();
   const feeAssets=new Set<string>();for(const f of g.fees){if(!f||!address(f.asset)||feeAssets.has(f.asset)||!count(f.decimals)||f.decimals>18||!amount(f.feeRaw)||!amount(f.taxRaw))fail();const prior=decimals.get(f.asset);if(prior!==undefined&&prior!==f.decimals)fail();decimals.set(f.asset,f.decimals);feeAssets.add(f.asset);}
  }
  const next=JSON.stringify(keys.sort());if(identity!==null&&identity!==next)fail();identity=next;
 }
 return p;
}
